### PR TITLE
Build proc-macro-test-impl out-of-tree

### PR DESCRIPTION
**Pull request by fasterthanlime**
Building it in-place fails in rust CI because the source directory is read-only. This changes `proc-macro-test`'s build script to first
copy `imp` under `OUT_DIR` (which is read-write).

It also prints stdout/stderr for the nested cargo invocation, should it fail. (I've seen failures in rust CI that I couldn't explain, and
when they take 25 minutes to reproduce, you want to have that info)

This change is tracked in:

  * https://github.com/rust-lang/rust-analyzer/issues/12818

Maintainer impact: none.

### CODE REVIEW

File: crates/proc-macro-test/build.rs [hunk 1]
```diff
@@ -12,14 +12,43 @@ use std::{
...
 
     let name = "proc-macro-test-impl";
     let version = "0.0.0";
+
+    let imp_dir = std::env::current_dir().unwrap().join("imp");
+
+    let staging_dir = out_dir.join("proc-macro-test-imp-staging");
+    // this'll error out if the staging dir didn't previously. using
+    // `std::fs::exists` would suffer from TOCTOU so just do our best to
+    // wip it and ignore errors.
```
Comment (lnicola) [inline crates/proc-macro-test/build.rs:28]:
```suggestion
    // wipe it and ignore errors.
```

Is `std::fs::remove_dir_all` still buggy? https://github.com/XAMPPRocky/remove_dir_all

File: crates/proc-macro-test/build.rs [hunk 2]
```diff
@@ -12,14 +12,43 @@ use std::{
...
     let out_dir = env::var_os("OUT_DIR").unwrap();
     let out_dir = Path::new(&out_dir);
 
     let name = "proc-macro-test-impl";
     let version = "0.0.0";
+
+    let imp_dir = std::env::current_dir().unwrap().join("imp");
+
+    let staging_dir = out_dir.join("proc-macro-test-imp-staging");
+    // this'll error out if the staging dir didn't previously. using
```
Comment (lnicola) [inline crates/proc-macro-test/build.rs:26]:
```suggestion
    // this'll error out if the staging dir didn't previously exist. using
```

File: crates/proc-macro-test/build.rs (hunk 1 above)
Comment (fasterthanlime) [reply crates/proc-macro-test/build.rs]: Actually my comment might be nonsense 🤔 I forget which of `remove_dir_all` / `create_dir_all` has this annoying edge case I keep running into in build scripts..

Anyway if this somehow fails the build we'll get a `panic!` later on in the build script. I can remove the comment / unwrap instead if you prefer.

Comment (fasterthanlime) [reply crates/proc-macro-test/build.rs]: Nope my instinct served me well, we can't unwrap this:

image

**Comment by fasterthanlime** [comment]
@Veykril this needs another r+, @lnicola caught some typos in comments and other commits snuck in

**Comment by fasterthanlime** [comment]
> In case you remove the temporary directory at the end (you might also need to `chdir` to the previous directory if you do that).

It's under `OUT_DIR`, so `cargo clean` will take care of it. We also overwrite it on every build where `proc-macro-test-impl` changes. It's also very small.

I'm gonna invoke "we might want to investigate that dir after a build" to not have to add additional cleanups 😄 

(If it was under `std::env::temp_dir()` we definitely would want to clean it up, that would add up)
